contracts/peripheral/YieldDirector.sol

```diff
@@ -0,0 +1,364 @@
+                    carry: 0,
+                    indexAtLastChange: index
+                })
+            );
+        } else {
+            DonationInfo storage donation = donations[uint256(recipientIndex)];
+
+            // Only update carry if there was a previous deposit
+            if(donation.deposit != 0) {
+                donation.carry += _getAccumulatedValue(donation.agnosticDeposit, donation.indexAtLastChange);
+            }
+
+            donation.deposit += amount_;
+            donation.agnosticDeposit = _toAgnostic(donation.deposit);
+            donation.indexAtLastChange = index;
+        }
+
+        RecipientInfo storage recipient = recipientInfo[recipient_];
+
+        // Calculate value carried over since last change
+        recipient.carry += _getAccumulatedValue(recipient.agnosticDebt, recipient.indexAtLastChange);
+        recipient.totalDebt += amount_;
+        recipient.agnosticDebt = _toAgnostic(recipient.totalDebt + recipient.carry);
+        recipient.indexAtLastChange = index;
+
+        emit Deposited(msg.sender, recipient_, amount_);
+    }
+
+
+    /**
+        @notice Withdraw donor's sOHM from vault and subtracts debt from recipient
+     */
+    function withdraw(uint256 amount_, address recipient_) external override {
+        require(disableWithdaws == false, "Withdraws currently disabled");
+
+        int256 recipientIndexSigned = _getRecipientIndex(msg.sender, recipient_);
+        require(recipientIndexSigned >= 0, "No donations to recipient");
+
+        uint256 index = IsOHM(sOHM).index();
+
+        // Donor accounting
+        DonationInfo storage donation = donationInfo[msg.sender][uint256(recipientIndexSigned)];
+
+        require(donation.deposit >= amount_, "Not enough sOHM to withdraw");
+
+        donation.carry += _getAccumulatedValue(donation.agnosticDeposit, donation.indexAtLastChange);
+        donation.deposit -= amount_;
+        donation.agnosticDeposit = _toAgnostic(donation.deposit);
+        donation.indexAtLastChange = index;
+
+        // Recipient accounting
+        RecipientInfo storage recipient = recipientInfo[recipient_];
+        recipient.carry += _getAccumulatedValue(recipient.agnosticDebt, recipient.indexAtLastChange);
+        recipient.totalDebt -= amount_;
+        recipient.agnosticDebt = _toAgnostic(recipient.totalDebt + recipient.carry);
+        recipient.indexAtLastChange = index;
+
+        IERC20(sOHM).safeTransfer(msg.sender, amount_);
+
+        emit Withdrawn(msg.sender, recipient_, amount_);
+    }
+
+    /**
+        @notice Withdraw from all donor positions
+     */
+    function withdrawAll() external override {
+        require(disableWithdaws == false, "Withdraws currently disabled");
+
+        DonationInfo[] storage donations = donationInfo[msg.sender];
+        require(donations.length != 0, "User not donating to anything");
+
+        uint256 sOhmIndex = IsOHM(sOHM).index();
+        uint256 total = 0;
```

Comment:
can just do `uint256 total;`

Reply:
Old habits 😅 I will keep though, I like the verbosity.